raft/raft.go

```diff
@@ -0,0 +1,403 @@
+package raft
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/gatewayd-io/gatewayd/config"
+	pb "github.com/gatewayd-io/gatewayd/raft/proto"
+	"github.com/hashicorp/raft"
+	raftboltdb "github.com/hashicorp/raft-boltdb"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+)
+
+// Command types for Raft operations.
+const (
+	CommandAddConsistentHashEntry = "ADD_CONSISTENT_HASH_ENTRY"
+	RaftLeaderState               = raft.Leader
+	LeaderElectionTimeout         = 3 * time.Second
+	maxSnapshots                  = 3                // Maximum number of snapshots to retain
+	maxPool                       = 3                // Maximum number of connections to pool
+	transportTimeout              = 10 * time.Second // Timeout for transport operations
+	leadershipCheckInterval       = 10 * time.Second // Interval for checking leadership status
+)
+
+// ConsistentHashCommand represents a command to modify the consistent hash.
+type ConsistentHashCommand struct {
+	Type      string `json:"type"`
+	Hash      uint64 `json:"hash"`
+	BlockName string `json:"blockName"`
+}
+
+// Node represents a node in the Raft cluster.
+type Node struct {
+	raft          *raft.Raft
+	config        *raft.Config
+	Fsm           *FSM
+	logStore      raft.LogStore
+	stableStore   raft.StableStore
+	snapshotStore raft.SnapshotStore
+	transport     raft.Transport
+	Logger        zerolog.Logger
+	Peers         []config.RaftPeer
+	rpcServer     *grpc.Server
+	rpcClient     *rpcClient
+	grpcAddr      string
+}
+
+// NewRaftNode creates and initializes a new Raft node.
+func NewRaftNode(logger zerolog.Logger, raftConfig config.Raft) (*Node, error) {
+	config := raft.DefaultConfig()
+
+	var err error
+	nodeID := raftConfig.NodeID
+	if raftConfig.NodeID == "" {
+		nodeID, err = os.Hostname()
+		if err != nil {
+			return nil, fmt.Errorf("error getting hostname: %w", err)
```

Comment:
I think we should wrap all errors with [GatewayDError](https://github.com/gatewayd-io/gatewayd/blob/main/errors/gatewayd_error.go).

Reply:
Addressed in [d6f286b](https://github.com/gatewayd-io/gatewayd/pull/636/commits/d6f286b177eedc663e1fca7dda76fcb8c957f778)

Reply:
I think we should avoid using [GatewayDError](https://github.com/gatewayd-io/gatewayd/blob/main/errors/gatewayd_error.go) in this case (in inner functions). Using it would prevent us from using the `errors` package in test cases, which has already caused challenges with other functions. Since this function fails with a fatal error, adding [GatewayDError](https://github.com/gatewayd-io/gatewayd/blob/main/errors/gatewayd_error.go) doesn’t seem to provide additional value. what do you think?